torch_em/self_training/probabilistic_unet_trainer.py

```diff
@@ -0,0 +1,116 @@
+import time
+import torch
+import torch_em
+
+
+class DummyLoss(torch.nn.Module):
+    pass
+
+
+class ProbabilisticUNetTrainer(torch_em.trainer.DefaultTrainer):
+    """This trainer implements training for the 'Probabilistic UNet' of Kohl et al.: (https://arxiv.org/abs/1806.05034).
+    This approach combines the learnings from UNet and VAEs (Prior and Posterior networks) to obtain generative
+    segmentations. The heuristic trains by taking into account the feature maps from UNet and the samples from
+    the posterior distribution, estimating the loss and further sampling from the prior for validation.
+
+    Parameters:
+        clip_posterior [bool] - (default: False)
+        prior_samples [int] - (default: 16)
+        supervised_loss [callable] - (default: None)
+        supervised_loss_and_metric [callable] - (default: None)
+    """
+
+    def __init__(
+            self,
+            clip_posterior=False,
+            prior_samples=16,
+            supervised_loss=None,
+            supervised_loss_and_metric=None,
+            **kwargs
+    ):
+        super().__init__(**kwargs)
+        assert supervised_loss, supervised_loss_and_metric is not None
+
+        self.supervised_loss = supervised_loss
+        self.supervised_loss_and_metric = supervised_loss_and_metric
```

Comment:
just a minor point, but we can just call it `loss` and `loss_and_metric` here, since this trainer is only for supervised training, so we don't need to make a distinction between the supervised and unsupervised cases.

Reply:
Yes, this makes sense. Updated it in the traning scheme as well as the loss. Thanks!